### PR TITLE
GI#94: Service to support app reviews on apps stores

### PR DESCRIPTION
**Pull request by ElNinjaGaiden**
You need to run this in order to install/sync new cordova dependencies:
```
npm install
npx cap sync
```

This PR prompts to users if they want to provide a review for `Beep` on their respective apps store once they have done the very first account/password check using the app.

### Rules
1. The prompt will be displayed only after the very first account/password check every time they use the app
1. If the user say he/she doesn't want to provide a review right now, but they keep using Beep to do more account/password checks, the prompt won't be displayed after this new checks
1. Once the user clicks on "Sure" (lets do the review), a flag will be stored into the Native storage of the device in order to avoid the prompt to be displayed anymore

### NOTES
In order to test the whole workflow, run the app on real devices. Simulators (at least iOS simulators) are not able to resolve app url's, they always return an error saying the url is not valid.

beepappreview

### CODE REVIEW

File: src/views/Pwd.vue [hunk 1]
```diff
@@ -120,6 +121,7 @@ export default {
         .get(this.getURL(hash.substr(0, 5)))
         .then(res => {
           const count = this.search(hash.substr(5).toUpperCase(), res.data)
+          this.$reviewAppService.registerReview()
```
[the remark this answers is not in the document]
Comment (ElNinjaGaiden) [reply src/views/Pwd.vue]: This is not asking the review right away, `registerReview` only increments the checks counter which will be used later when navigation happens

[the remark this answers is not in the document]
Comment (ElNinjaGaiden) [reply src/views/Pwd.vue]: Maybe a better name for that function will be `registerCheck`

File: src/reviewAppService.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      ((this.revisionsDone === 1 && !this.storeReviewDone && !this.storeReviewRejected) ||
+        forceReview)
+    ) {
+      this.doPromptAppReview()
+    }
+  },
+  doPromptAppReview() {
+    Modals.confirm({
+      title: 'Have some feedback?',
+      message: 'Do you want to review this app?',
```
Comment (grgur) [inline src/reviewAppService.js]: Can we remove the title and just ask
> Did you find BEEP helpful?

Yes -> Review
No -> discard

I'd actually also consider doing something else...

1. Is it possible to convert modal to a full size view?
2. Can we ask two questions:
_Did you like using BEEP?_

if *yes*: _Would you mind rating BEEP?_ (yes takes us to the app store rating page)

if *no*: _We would love to learn from your experience. Can you give us feedback, please?_ - then show a textarea that gets sent to our e-mail.

**Comment by ElNinjaGaiden** [comment]
@grgur requested changes are done. The only question remaining is what will be the email address to use for the feedback??

CC @michaeltintiuc

**Comment by ElNinjaGaiden** [comment]
Sure @grgur CC @michaeltintiuc 

### Workflow on web environment (only feedback via email displayed):
appreviewweb

### Workflow Native 1 (select app review on app store):
appreviewandroid1

### Workflow Native 2 (select feedback my email):
appreviewandroid2

**Comment by ElNinjaGaiden** [comment]
@grgur we could also ask to the design team for some kind of asset to give some love to this _provide feedback_ view. Right now it looks so bla.

**Comment by falkonprods** [comment]
@grgur I recommend we proceed with your plan to offer a popup to either (a) positively rate, or (b) send feedback to Modus. That is a best practice.

**Comment by tasteofdalife** [comment]
@ElNinjaGaiden @grgur @falkonprods Hey, everyone, while I'm working on the design, please consider showing this flow not after 1st use of the app, but after 3rd or even 5th, since it's the best practice. After the 1st use user may still not know or like the app enough to give us some useful feedback and moreover, actually like the app. He may perceive this ask as annoying and even leave some negative comment. Also, regardless of his answer (unless it's "Maybe Later"), we should not show this message again. However we should show it again upon some major app update (because reviews are shown on the app store based on app version, not on the overall app).

**Comment by tasteofdalife** [comment]
@ElNinjaGaiden @grgur is there some particular reason why we're not using native OS system popups to ask this question as oppose to the full screen view?

**Comment by ElNinjaGaiden** [comment]
@tasteofdalife the initial implementation was using native popups but they are kind of small and don't provide enough space/nice experience imo (and I would guess @grgur agree since he notice the problem first). Some of the text were getting cut off.

On the other hand, all your recommendations regarding when to ask for feedback make totally sense to me.

**Comment by michaeltintiuc** [comment]
@ElNinjaGaiden  @tasteofdalife @grgur Capacitor provides native popups which are diferent from Ionic's we can see how they look, also we could use Ionic's modals which are fullscreen

**Comment by ElNinjaGaiden** [comment]
@michaeltintiuc Capacitor popups were used on the initial implementation (see first attached .gif) then @grgur suggested to change to full screen modals. We are already using ionic modals for this feature.

**Comment by grgur** [comment]
@tasteofdalife I agree with timing.

@tasteofdalife @ElNinjaGaiden @michaeltintiuc My whole goal was not to ask for a review right away. I'd like only the satisfied users to be prompted with the review dialog. To do that, we need to add an extra step/question. I'd like to ask users if they are happy with it before we give them the tool to review us.

How users review is up to the team to decide. I'm all for using the native dialog.

**Comment by ElNinjaGaiden** [comment]
@grgur I'm confused with this new idea/proposed flow. The idea is to use native pop ups (or whatever) to ask users if they are happy with beep?

If the answer is yes, then show the current fullscreen view we have right now (after apply UI changes made by @tasteofdalife) in order to let users to rate beep or to send us en email with feedback?

If the answer is no, do nothing or ask them why they are not happy with it, via email probably? (I'm trying to understand the flow)

**Comment by grgur** [comment]
@ElNinjaGaiden Great questions. I think @tasteofdalife's designs explain it better than words

## Step 1 - Use a custom view to ask if they like the app
image

## Step 2 - If they say they are happy, let them rate via the native dialog
image

If they were not happy, then we need another screen where they can give us private feedback.

Does that help?

**Comment by ElNinjaGaiden** [comment]
Bump! @grgur @michaeltintiuc @tasteofdalife

**Comment by ElNinjaGaiden** [comment]
@michaeltintiuc @grgur conflicts resolved

**Comment by michaeltintiuc** [comment]
@ElNinjaGaiden please sync with master to fix the conflicts, tested on Android 9, seems good.
The workflow is somewhat confusing for me though, the review prompt wasn't shown after the 1st password check, could you please confirm the steps? Also if we update, I believe the code is resetting the counter right?

**Comment by grgur** [comment]
Thank you @ElNinjaGaiden!!!!!

**Comment by grgur** [comment]
Merged manually so i'm closing here

**Pull request closed without merging**